src/env.c

```diff
@@ -0,0 +1,57 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <stdlib.h>
+
+#include "rcutils/env.h"
+#include "rcutils/error_handling.h"
+
+// TODO(cottsay): Move the stuff in get_env.c in here
+
+bool
+rcutils_set_env(const char * env_name, const char * env_value)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    env_name, "env_name is null", return false);
+
+#ifdef _WIN32
+  if (NULL == env_value) {
+    env_value = "";
+  }
+  if (0 != _putenv_s(env_name, env_value)) {
```

Comment:
For Unix, we are careful to use `setenv` as it makes copies of the strings.  From https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/putenv-s-wputenv-s?view=vs-2019 , it isn't clear to me whether Windows does the same.  Do you know whether it makes copies or not?

Reply:
I assume that `setenv` is calling `malloc` then. Even `putenv`, which doesn't duplicate the string, can return `ENOMEM`, so I would assume that it is (re-)allocating memory as well. Is that a problem, or would documentation be sufficient? Are you aware of another way to do this where we can use an allocator?

I'll look into answering the question for Windows.

Reply:
I found some MSDN docs discussing the way environment variables are implemented in Windows: https://docs.microsoft.com/en-us/cpp/c-runtime-library/environ-wenviron?view=vs-2019

It is similar to Linux in that the key and value are stored in the same string, so calling a function that takes them separately can't re-use the strings and must be allocating new memory to store them combined.

I also tested whether the `_putenv`, which takes a single string like the Linux `putenv` function does, is allocating new memory, and it seems to be doing so and copying the string passed to `_putenv`. That said, I can't find any documentation explicitly stating that it happens like that all of the the time.

tl;dr - For our purposes, `_putenv_s` should act similar to `setenv` as far as ownership of the strings passed into it.

Reply:
> I assume that `setenv` is calling `malloc` then.

Yeah, I would guess so.  The man page for it says:

```
This function makes copies of the strings pointed to by name and value (by contrast with putenv(3)).
```

> Even `putenv`, which doesn't duplicate the string, can return `ENOMEM`, so I would assume that it is (re-)allocating memory as well.

My reading of glibc here is that `putenv` can indeed return ENOMEM.  In particular, this can happen if we are putting a *new* value into the environment, which can cause the environment to grow, which requires a realloc.  A replacement of an existing variable doesn't do any allocation at all for `putenv`.  For a new environment variable, `setenv` does two allocations: one to expand the environment and one to copy the string given by the user into glibc-controlled memory.  For a replacement, `setenv` does a single allocation to copy the string given by the user into glibc-controlled memory.

In any case, I don't think there is much we can do about that allocation; it happens deep in glibc.  What I'm aiming for here is consistency between platforms, which is where the question about Windows comes from.  Once we know the answer to that, we'll have a better idea on how to proceed.

Reply:
> tl;dr - For our purposes, `_putenv_s` should act similar to `setenv` as far as ownership of the strings passed into it.

Yeah, I just did a quick test here on Windows, and it does indeed seem like `_putenv_s` makes a copy of the environment string (at least, modifying it in place didn't change the returned value from getenv).  So I think we are good to go with this; just documentation that this function always *copies* into the environment is good enough for now.

Reply:
Done in b697ce0